### PR TITLE
support for verge

### PR DESCRIPTION
**Pull request by justinvforvendetta**
please feel free to reach out to me if there's anything else i need to do or edit
justinvforvendetta@gmail.com

### CODE REVIEW

**Comment by marpme** [comment]
Looks good to me — ACK

**Comment by martinboehm** [comment]
There are no tests in your PR. Please see the content of some accepted PRs for other coins as an example.

**Comment by justinvforvendetta** [comment]
@martinboehm is https://github.com/trezor/blockbook/pull/318/commits/0bd08ac64093d974f2d0044dbd44fa3c1f727085#diff-957e27e9408fed500cdfde6ab579202fR89 just the hex of the block itself? 

and what value is testTxPacked1 looking for?

**Comment by justinvforvendetta** [comment]
@martinboehm if you get a moment, let me know how it looks, or you can drop me a line justinvforvendetta@gmail.com

**Comment by martinboehm** [comment]
@justinvforvendetta Hi, the unit test failed:
```
# blockbook/bchain/coins/verge
bchain/coins/verge/vergeparser_test.go:125:3: missing ',' before newline in composite literal
bchain/coins/verge/vergeparser_test.go:128:6: expected '(', found Test_PackTx
bchain/coins/verge/vergeparser_test.go:168:6: expected '(', found Test_UnpackTx
bchain/coins/verge/vergeparser_test.go:207:2: expected '}', found 'EOF'
FAIL	blockbook/bchain/coins/verge [setup failed]
```
and the integration test is missing.

**Comment by justinvforvendetta** [comment]
@martinboehm sorry to bother you again, but i'm having an issue getting the first piece of the testTxPacked1 value. i see the remainder of the string is the same as the hex, and im guessing the first part is related to the value? if you could clarify that for me, it would be greatly appreciated.

**Comment by justinvforvendetta** [comment]
hey @martinboehm i seem to be stuck, im fairly certain i've done the parser test correctly, but i'm getting an unexpected end of file error, on line 191. https://github.com/trezor/blockbook/pull/318/files#diff-957e27e9408fed500cdfde6ab579202fR191

File: tests/rpc/testdata/verge.json [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+                }
+
+            ],
+            "vout": [
+                {
+                    "value": 17230.000000,
+                    "n": 0,
+                    "scriptPubKey": {
+                        "hex": "76a91431ae5739fbb60574e10323f38538ada2730826c388ac"
+                    }get
```
Comment (wakiyamap) [inline tests/rpc/testdata/verge.json]: get? Is this correct?

Comment (justinvforvendetta) [reply tests/rpc/testdata/verge.json]: @wakiyamap no, but doesnt fix the issue im experiencing =\

**Comment by justinvforvendetta** [comment]
@martinboehm ok i think we got it now, ready for your review =]

**Comment by martinboehm** [comment]
Hi, I am able to build both the backend and blockbook packages. However, when I install the verge backend and try to run it, I get
```
verged[15911]: /opt/coins/nodes/verge/bin/verged: /lib/x86_64-linux-gnu/libm.so.6: version `GLIBC_2.27' not found (required by /opt/coins/nodes/verge/bin/verged)
```
I run it on Debian GNU/Linux 9 (stretch)

**Comment by justinvforvendetta** [comment]
do you have an earlier version installed already? 
http://ftp.gnu.org/gnu/glibc/glibc-2.27.tar.gz

**Comment by martinboehm** [comment]
I am running it on Debian 9, which has GLIBC 2.24. It seems that Verge is not compatible with Debian 9.
As I don't want to mess with our server, could you please make the Verge backend RPC interface available somewhere? I would run only Verge Blockbook and test it against your backend.

**Comment by justinvforvendetta** [comment]
i will set one up for you, whats your email? ill deploy one tonight and send you the ip and creds

**Comment by martinboehm** [comment]
Send it please to martin.bohm@satoshilabs.com. Thanks.

**Comment by justinvforvendetta** [comment]
@martinboehm sent, thanks!

**Comment by atomlab** [comment]
@justinvforvendetta Hello. I have [forked](https://github.com/Atomicwallet/blockbook/tree/verge) code from this PR and builded blockbook with verge changes on Debian10. I have some porblem. 

`./blockbook/logs/blockbook.WARNING`
```
Log file created at: 2020/04/12 09:10:05
Running on machine: xvg-03
Binary: Built with gc go1.12.4 for linux/amd64
Log line format: [IWEF]mmdd hh:mm:ss.uuuuuu threadid file:line] msg
E0412 09:10:05.773256   21488 sync.go:302] getBlockWorker 1 connect block error 2 00000ce17211c385317243edea7432ebe7284804ba880d0d2c8b05f1e8c22e28: readScript: transaction input signature script is larger than the max allowed size [count 432628491087773695, max 1073741824]. Retrying...
E0412 09:10:05.774127   21488 sync.go:302] getBlockWorker 0 connect block error 1 000007342680f4a3d7a3d34a9214b5b71c6f1bec3d633d6d2b84d474a6be55fd: readScript: transaction input signature script is larger than the max allowed size [count 432628486792806399, max 1073741824]. Retrying...
E0412 09:10:05.775032   21488 sync.go:302] getBlockWorker 4 connect block error hash 00000fc63692467faeb20cdb3b53200dc601d75bdfa1001463304cc790d77278: readScript: transaction input signature script is larger than the max allowed size [count 72056512175931391, max 1073741824]. Retrying...
E0412 09:10:05.776121   21488 sync.go:302] getBlockWorker 2 connect block error 3 00000e79cfec909e108e7073c7b313854b22e9850a2fbdd475db31096a044889: readScript: transaction input signature script is larger than the max allowed size [count 432628495382740991, max 1073741824]. Retrying...
E0412 09:10:05.776972   21488 sync.go:302] getBlockWorker 3 connect block error 4 00000f46c405eb645a265ccc627730fdf65ff273467239485a80db9ac09448aa: readScript: transaction input signature script is larger than the max allowed size [count 433472924607840255, max 1073741824]. Retrying...
E0412 09:10:05.779873   21488 sync.go:302] getBlockWorker 7 connect block error 5 00000e501c6bfe314d3dcf7e4994d1416dda45dc9c49d6ad5d3e730a92abc47c: readScript: transaction input signature script is larger than the max allowed size [count 5777843448326720255, max 1073741824]. Retrying...
E0412 09:10:05.780296   21488 sync.go:302] getBlockWorker 5 connect block error 6 000007bf9d147bb50fd5fffa08382a0ef97438dcfbf2e1076a2b5582ed006213: readScript: transaction input signature script is larger than the max allowed size [count 432909983244353535, max 1073741824]. Retrying...
E0412 09:10:05.780377   21488 sync.go:302] getBlockWorker 6 connect block error 7 000006bb9a994742bda6b904c1e7162b61b87a531b1e4a7abed3ab855c844284: readScript: transaction input signature script is larger than the max allowed size [count 433191462516031487, max 1073741824]. Retrying...
E0412 09:10:06.277925   21488 sync.go:302] getBlockWorker 1 connect block error 2 00000ce17211c385317243edea7432ebe7284804ba880d0d2c8b05f1e8c22e28: readScript: transaction input signature script is larger than the max allowed size [count 432628491087773695, max 1073741824]. Retrying...
E0412 09:10:06.280151   21488 sync.go:302] getBlockWorker 0 connect block error 1 000007342680f4a3d7a3d34a9214b5b71c6f1bec3d633d6d2b84d474a6be55fd: readScript: transaction input signature script is larger than the max allowed size [count 432628486792806399, max 1073741824]. Retrying...
E0412 09:10:06.282759   21488 sync.go:302] getBlockWorker 2 connect block error 3 00000e79cfec909e108e7073c7b313854b22e9850a2fbdd475db31096a044889: readScript: transaction input signature script is larger than the max allowed size [count 432628495382740991, max 1073741824]. Retrying...
E0412 09:10:06.285352   21488 sync.go:302] getBlockWorker 3 connect block error 4 00000f46c405eb645a265ccc627730fdf65ff273467239485a80db9ac09448aa: readScript: transaction input signature script is larger than the max allowed size [count 433472924607840255, max 1073741824]. Retrying...
E0412 09:10:06.288553   21488 sync.go:302] getBlockWorker 7 connect block error 5 00000e501c6bfe314d3dcf7e4994d1416dda45dc9c49d6ad5d3e730a92abc47c: readScript: transaction input signature script is larger than the max allowed size [count 5777843448326720255, max 1073741824]. Retrying...
E0412 09:10:06.290441   21488 sync.go:302] getBlockWorker 4 connect block error hash 00000fc63692467faeb20cdb3b53200dc601d75bdfa1001463304cc790d77278: readScript: transaction input signature script is larger than the max allowed size [count 72056512175931391, max 1073741824]. Retrying...
E0412 09:10:06.292870   21488 sync.go:302] getBlockWorker 5 connect block error 6 000007bf9d147bb50fd5fffa08382a0ef97438dcfbf2e1076a2b5582ed006213: readScript: transaction input signature script is larger than the max allowed size [count 432909983244353535, max 1073741824]. Retrying...
E0412 09:10:06.295399   21488 sync.go:302] getBlockWorker 6 connect block error 7 000006bb9a994742bda6b904c1e7162b61b87a531b1e4a7abed3ab855c844284: readScript: transaction input signature script is larger than the max allowed size [count 433191462516031487, max 1073741824]. Retrying...
E0412 09:10:06.781781   21488 sync.go:302] getBlockWorker 1 connect block error 2 00000ce17211c385317243edea7432ebe7284804ba880d0d2c8b05f1e8c22e28: readScript: transaction input signature script is larger than the max allowed size [count 432628491087773695, max 1073741824]. Retrying...
E0412 09:10:06.784198   21488 sync.go:302] getBlockWorker 0 connect block error 1 000007342680f4a3d7a3d34a9214b5b71c6f1bec3d633d6d2b84d474a6be55fd: readScript: transaction input signature script is larger than the max allowed size [count 432628486792806399, max 1073741824]. Retrying...
```

verge node verstion 

```
Verge Leviosa Daemon version v6.0.0.0
```

`verged.conf`

```

datadir = /opt/verge/storage
daemon = 1
server = 1
nolisten = 1
rpcuser = rpc
rpcpassword = rpc
rpcport = 20102
port = 21102
txindex = 1
without-tor = 1

zmqpubhashtx = tcp://127.0.0.1:38346
zmqpubhashblock = tcp://127.0.0.1:38346

rpcworkqueue = 1100
maxmempool = 2000
dbcache = 1000
whitelist = 127.0.0.1
```

verge `getblockchaininfo`
>{"result":{"chain":"main","blocks":3973934,"headers":3973934,"bestblockhash":"564333db2edb4fd3871dc2402f29ca427beaf40735d74df9b6a53429e46503ee","difficulty":16709.9244826419,"mediantime":1586682504,"verificationprogress":0.9999890199414572,"initialblockdownload":false,"chainwork":"0000000000000000000000000000000000000000004d8e2c26a439ab0eadef9d","size_on_disk":6505291868,"pruned":false,"softforks":[{"id":"bip34","version":2,"reject":{"status":true}},{"id":"bip66","version":3,"reject":{"status":true}},{"id":"bip65","version":4,"reject":{"status":true}}],"bip9_softforks":{"csv":{"status":"failed","startTime":1529247969,"timeout":1559347200,"since":3173000},"segwit":{"status":"failed","startTime":1529247969,"timeout":1559347200,"since":3173000}},"warnings":""},"error":null,"id":1}


`blockchaincfg.json`
```
{
    "address_format": "",
    "block_addresses_to_keep": 300,
    "coin_label": "Verge",
    "coin_name": "Verge",
    "coin_shortcut": "XVG",
    "mempool_sub_workers": 2,
    "mempool_workers": 8,
    "message_queue_binding": "tcp://127.0.0.1:38346",
    "parse": true,
    "rpc_pass": "rpc",
    "rpc_timeout": 25,
    "rpc_url": "http://127.0.0.1:20102",
    "rpc_user": "rpc",
    "slip44": 77,
    "subversion": "",
    "xpub_magic": 36513075,
    "xpub_magic_segwit_native": 78792518,
    "xpub_magic_segwit_p2sh": 77429938
}
```

`curl http://localhost:9188/api/`
>{"blockbook":{"coin":"Verge","host":"xvg-03","version":"devel","gitCommit":"4747c05","buildTime":"2020-04-10T18:27:30+00:00","syncMode":true,"initialSync":true,"inSync":false,"bestHeight":0,"lastBlockTime":"0001-01-01T00:00:00Z","inSyncMempool":false,"lastMempoolTime":"0001-01-01T00:00:00Z","mempoolSize":0,"decimals":8,"dbSize":166367,"about":"Blockbook - blockchain indexer for Trezor wallet https://trezor.io/. Do not use for any other purpose."},"backend":{"chain":"main","blocks":3973941,"headers":3973941,"bestBlockHash":"b479f821ba955ee20095d04fdc040b3dc5b64266a0247e890ce8723421585702","difficulty":"141482652.240816","sizeOnDisk":6505294775,"version":"6000000","subversion":"/verge:6.0.0/","protocolVersion":"90008"}}
```

**Comment by justinvforvendetta** [comment]
@atomlab what about 6.0.2 please

**Comment by atomlab** [comment]
> @atomlab what about 6.0.2 please, also is verged running with a transaction index? (txindex=1 in .conf?)

After update verge up to 6.0.2 blockbook start and crash with error

```
$ rm -rf /opt/blockbook/data/db
$ /opt/blockbook/bin/blockbook -blockchaincfg=/opt/blockbook/etc/blockchaincfg.json -datadir=/opt/blockbook/data/db -sync -internal=:9088 -public=:9188 -explorer= -log_dir=/opt/blockbook/logs -dbcache=0 -workers=1

E0413 08:33:51.268871    1647 blockbook.go:291] resyncIndex hash 00000fc63692467faeb20cdb3b53200dc601d75bdfa1001463304cc790d77278: readScript: transaction input signature script is larger than the max allowed size [count 72056512175931391, max 1073741824]
$ echo $?
255
$
```

```
$ verged --version
Verge Leviosa Daemon version v6.0.2.0
```

verged.conf
```
datadir = /opt/verge/storage
daemon = 1
server = 1
rpcuser = rpc
rpcpassword = rpc
rpcport = 20102
port = 21102
txindex = 1
without-tor = 1
rpcworkqueue = 1100
maxmempool = 2000
dbcache = 1000
whitelist = 127.0.0.1
```

```
curl -s -H 'Content-type:text/plain' -X POST -d '{"jsonrpc": 2.0, "id": 1, "method": "getblockchaininfo", "params": []}' http://rpc:rpc@localhost:20102
```

>{"result":{"chain":"main","blocks":3976360,"headers":3976360,"bestblockhash":"00000000000017caf29c34cdb2ac46828b18d27585e26df874b3b380b74619df","difficulty":5870.624282386904,"mediantime":1586767035,"verificationprogress":0.9999996655852542,"initialblockdownload":false,"chainwork":"0000000000000000000000000000000000000000004da73e9df0b60572b7d28d","size_on_disk":6508154012,"pruned":false,"softforks":[{"id":"bip34","version":2,"reject":{"status":true}},{"id":"bip66","version":3,"reject":{"status":true}},{"id":"bip65","version":4,"reject":{"status":true}}],"bip9_softforks":{"csv":{"status":"failed","startTime":1529247969,"timeout":1559347200,"since":3173000},"segwit":{"status":"failed","startTime":1529247969,"timeout":1559347200,"since":3173000}},"warnings":""},"error":null,"id":1}

**Comment by justinvforvendetta** [comment]
can you try this again with a fresh blockbook db?

**Comment by atomlab** [comment]
> can you try this again with a fresh blockbook db?

I have removed `blockbook/db` and restart blockbook again. 
Blockbook builded from this [branch](https://github.com/Atomicwallet/blockbook/tree/verge):
- branch contains yours [PR](https://github.com/Atomicwallet/blockbook/commit/4747c05ddd4862d218bf2e4d85940e2fca1748ab)
- branch also [merged](https://github.com/Atomicwallet/blockbook/commit/c2a50be9d6015e4f2fb71ce35d912bbe93d448a0) from master.

When I start blockbook I see an error and blockbook crash.
>May 02 08:25:15 xvg-03 systemd[1]: Started Blockbook daemon.
May 02 08:25:16 xvg-03 blockbook[30362]: E0502 08:25:16.268733   30362 blockbook.go:291] resyncIndex hash 00000fc63692467faeb20cdb3b53200dc601d75bdfa1001463304cc790d77278: readScript: transaction input signature script is larger than the max allowed size [count 72056512175931391, max 1073741824]
May 02 08:25:16 xvg-03 systemd[1]: blockbook.service: Main process exited, code=exited, status=255/EXCEPTION
May 02 08:25:16 xvg-03 systemd[1]: blockbook.service: Failed with result 'exit-code'.


logs 
```
#cat ./blockbook/logs/blockbook.ERROR
```
>Log file created at: 2020/05/02 08:39:52
Running on machine: xvg-03
Binary: Built with gc go1.12.4 for linux/amd64
Log line format: [IWEF]mmdd hh:mm:ss.uuuuuu threadid file:line] msg
E0502 08:39:52.197572    7040 blockbook.go:291] resyncIndex hash 00000fc63692467faeb20cdb3b53200dc601d75bdfa1001463304cc790d77278: readScript: transaction input signature script is larger than the max allowed size [count 72056512175931391, max 1073741824]

verge version 
```
# /opt/verge/bin/verged --version
Verge Leviosa Daemon version v6.0.2.0
```
verge last block
```
curl -s -H 'Content-type:text/plain' -X POST -d '{"jsonrpc": 2.0, "id": 1, "method": "getblockchaininfo", "params": []}' http://rpc:rpc@localhost:20102
```
> {"result":{"chain":"main","blocks":4023082,"headers":4023082,"bestblockhash":"7569a41f751b92b79f0f1bb7ffef07e79d2cf6519c6ddd9d7f74408ab7100bf7","difficulty":48480921.88855874,"mediantime":1588408023,"verificationprogress":0.9999999573246804,"initialblockdownload":false,"chainwork":"000000000000000000000000000000000000000000505c80bb1997e670c2eabb","size_on_disk":6545355105,"pruned":false,"softforks":[{"id":"bip34","version":2,"reject":{"status":true}},{"id":"bip66","version":3,"reject":{"status":true}},{"id":"bip65","version":4,"reject":{"status":true}}],"bip9_softforks":{"csv":{"status":"failed","startTime":1529247969,"timeout":1559347200,"since":3173000},"segwit":{"status":"failed","startTime":1529247969,"timeout":1559347200,"since":3173000}},"warnings":""},"error":null,"id":1}

**Comment by martinboehm** [comment]
Closed for inactivity.

**Pull request closed without merging**